File: admin-frontend/src/lib/permissions.ts
import { AdminRole } from '@/types';

/**
 * Permission definitions matching backend ROLE_PERMISSIONS
 * 
 * Sync with: admin-backend/app/utils/permissions.py
 */
type Permission =
  // Dashboard
  | 'view_dashboard'
  | 'view_metrics'
  // Users
  | 'view_users'
  | 'view_user_details'
  | 'ban_users'
  | 'adjust_balance'
  | 'approve_large_adjustment'
  // Rooms
  | 'view_rooms'
  | 'create_room'      // 방 생성 권한 추가
  | 'update_room'      // 방 수정 권한 추가
  | 'delete_room'      // 방 삭제 권한 추가
  | 'force_close_room'
  | 'send_room_message'
  // Hands
  | 'view_hands'
  | 'export_hands'
  // Crypto
  | 'view_deposits'
  | 'view_withdrawals'
  | 'approve_withdrawal'
  | 'approve_large_withdrawal'
  | 'view_wallet'
  // Announcements
  | 'view_announcements'
  | 'create_announcement'
  | 'broadcast_announcement'
  // Maintenance
  | 'view_maintenance'
  | 'schedule_maintenance'
  | 'activate_maintenance'
  // Suspicious
  | 'view_suspicious'
  | 'resolve_suspicious'
  // Audit
  | 'view_audit_logs'
  // Admin Management
  | 'view_admins'
  | 'create_admin'
  | 'modify_admin'
  // Partners
  | 'view_partners'
  | 'create_partner'
  | 'update_partner'
  | 'delete_partner'
  // Settlements
  | 'view_settlements'
  | 'generate_settlement'
  | 'approve_settlement'
  | 'pay_settlement';

const rolePermissions: Record<AdminRole, Permission[]> = {
  [AdminRole.VIEWER]: [
    'view_dashboard',
    'view_metrics',
    'view_users',
    'view_rooms',
    'view_hands',
    'view_deposits',
    'view_withdrawals',
    'view_wallet',
    'view_announcements',
    'view_maintenance',
    'view_suspicious',
    'view_partners',
    'view_settlements',
  ],
  [AdminRole.OPERATOR]: [
    // All viewer permissions
    'view_dashboard',
    'view_metrics',
    'view_users',
    'view_user_details',
    'view_rooms',
    'view_hands',
    'export_hands',
    'view_deposits',
    'view_withdrawals',
    'view_wallet',
    'view_announcements',
    'view_maintenance',
    'view_suspicious',
    'view_partners',
    'view_settlements',
    // Operator-specific
    'ban_users',
    'send_room_message',
    'approve_withdrawal',
    'create_announcement',
    'resolve_suspicious',
    'view_audit_logs',
  ],
  [AdminRole.SUPERVISOR]: [
    // All operator permissions
    'view_dashboard',
    'view_metrics',
    'view_users',
    'view_user_details',
    'ban_users',
    'view_rooms',
    'send_room_message',
    'view_hands',
    'export_hands',
    'view_deposits',
    'view_withdrawals',
    'approve_withdrawal',
    'view_wallet',
    'view_announcements',
    'create_announcement',
    'view_maintenance',
    'view_suspicious',
    'resolve_suspicious',
    'view_audit_logs',
    'view_partners',
    'view_settlements',
    // Supervisor-specific
    'adjust_balance',
    'approve_large_withdrawal',
    'create_room',        // 방 생성 권한
    'update_room',        // 방 수정 권한
    'force_close_room',
    'broadcast_announcement',
    'schedule_maintenance',
    'create_partner',
    'update_partner',
    'generate_settlement',
    'approve_settlement',
  ],
  [AdminRole.ADMIN]: [
    // All permissions
    'view_dashboard',
    'view_metrics',
    'view_users',
    'view_user_details',
    'ban_users',
    'adjust_balance',
    'approve_large_adjustment',
    'view_rooms',
    'create_room',        // 방 생성 권한
    'update_room',        // 방 수정 권한
    'delete_room',        // 방 삭제 권한
    'force_close_room',
    'send_room_message',
    'view_hands',
    'export_hands',
    'view_deposits',
    'view_withdrawals',
    'approve_withdrawal',
    'approve_large_withdrawal',
    'view_wallet',
    'view_announcements',
    'create_announcement',
    'broadcast_announcement',
    'view_maintenance',
    'schedule_maintenance',
    'activate_maintenance',
    'view_suspicious',
    'resolve_suspicious',
    'view_audit_logs',
    'view_admins',
    'create_admin',
    'modify_admin',
    // Partners
    'view_partners',
    'create_partner',
    'update_partner',
    'delete_partner',
    // Settlements
    'view_settlements',
    'generate_settlement',
    'approve_settlement',
    'pay_settlement',
  ],
  // Partner role - 자신의 데이터만 조회 가능 (파트너 포털 전용)
  [AdminRole.PARTNER]: [
    'view_dashboard',
    'view_settlements',
  ],
};

export function hasPermission(role: AdminRole, permission: Permission): boolean {
  return rolePermissions[role]?.includes(permission) ?? false;
}

export function canAccessRoute(role: AdminRole, route: string): boolean {
  // Route to permission mapping
  const routePermissions: Record<string, Permission> = {
    '/': 'view_dashboard',
    '/users': 'view_users',
    '/bans': 'ban_users',
    '/rooms': 'view_rooms',
    '/hands': 'view_hands',
    '/deposits': 'view_deposits',
    '/suspicious': 'view_suspicious',
    '/announcements': 'view_announcements',
    '/maintenance': 'view_maintenance',
    '/audit': 'view_audit_logs',
    '/settings': 'view_admins',
    '/partners': 'view_partners',
    '/settlements': 'view_settlements',
  };

  const permission = routePermissions[route];
  if (!permission) return true;

  return hasPermission(role, permission);
}

/**
 * Get all permissions for a role
 */
export function getRolePermissions(role: AdminRole): Permission[] {
  return rolePermissions[role] ?? [];
}

File: admin-backend/app/utils/permissions.py
from enum import Enum
from typing import Set
from app.models.admin_user import AdminRole


class Permission(str, Enum):
    # Dashboard
    VIEW_DASHBOARD = "view_dashboard"
    VIEW_METRICS = "view_metrics"

    # Users
    VIEW_USERS = "view_users"
    VIEW_USER_DETAILS = "view_user_details"
    BAN_USERS = "ban_users"
    ADJUST_BALANCE = "adjust_balance"
    APPROVE_LARGE_ADJUSTMENT = "approve_large_adjustment"

    # Rooms
    VIEW_ROOMS = "view_rooms"
    CREATE_ROOM = "create_room"
    UPDATE_ROOM = "update_room"
    DELETE_ROOM = "delete_room"
    FORCE_CLOSE_ROOM = "force_close_room"
    SEND_ROOM_MESSAGE = "send_room_message"

    # Hands
    VIEW_HANDS = "view_hands"
    EXPORT_HANDS = "export_hands"

    # Crypto
    VIEW_DEPOSITS = "view_deposits"
    VIEW_WITHDRAWALS = "view_withdrawals"
    APPROVE_WITHDRAWAL = "approve_withdrawal"
    APPROVE_LARGE_WITHDRAWAL = "approve_large_withdrawal"
    VIEW_WALLET = "view_wallet"

    # Announcements
    VIEW_ANNOUNCEMENTS = "view_announcements"
    CREATE_ANNOUNCEMENT = "create_announcement"
    BROADCAST_ANNOUNCEMENT = "broadcast_announcement"

    # Maintenance
    VIEW_MAINTENANCE = "view_maintenance"
    SCHEDULE_MAINTENANCE = "schedule_maintenance"
    ACTIVATE_MAINTENANCE = "activate_maintenance"

    # Suspicious
    VIEW_SUSPICIOUS = "view_suspicious"
    RESOLVE_SUSPICIOUS = "resolve_suspicious"

    # Audit
    VIEW_AUDIT_LOGS = "view_audit_logs"

    # Admin Management
    VIEW_ADMINS = "view_admins"
    CREATE_ADMIN = "create_admin"
    MODIFY_ADMIN = "modify_admin"

    # Partners (총판 관리)
    VIEW_PARTNERS = "view_partners"
    CREATE_PARTNER = "create_partner"
    UPDATE_PARTNER = "update_partner"
    DELETE_PARTNER = "delete_partner"

    # Settlements (정산)
    VIEW_SETTLEMENTS = "view_settlements"
    GENERATE_SETTLEMENT = "generate_settlement"
    APPROVE_SETTLEMENT = "approve_settlement"
    PAY_SETTLEMENT = "pay_settlement"


# Role to permissions mapping
ROLE_PERMISSIONS: dict[AdminRole, Set[Permission]] = {
    AdminRole.viewer: {
        Permission.VIEW_DASHBOARD,
        Permission.VIEW_METRICS,
        Permission.VIEW_USERS,
        Permission.VIEW_ROOMS,
        Permission.VIEW_HANDS,
        Permission.VIEW_DEPOSITS,
        Permission.VIEW_WITHDRAWALS,
        Permission.VIEW_WALLET,
        Permission.VIEW_ANNOUNCEMENTS,
        Permission.VIEW_MAINTENANCE,
        Permission.VIEW_SUSPICIOUS,
    },
    AdminRole.operator: {
        # All viewer permissions
        Permission.VIEW_DASHBOARD,
        Permission.VIEW_METRICS,
        Permission.VIEW_USERS,
        Permission.VIEW_USER_DETAILS,
        Permission.VIEW_ROOMS,
        Permission.VIEW_HANDS,
        Permission.EXPORT_HANDS,
        Permission.VIEW_DEPOSITS,
        Permission.VIEW_WITHDRAWALS,
        Permission.VIEW_WALLET,
        Permission.VIEW_ANNOUNCEMENTS,
        Permission.VIEW_MAINTENANCE,
        Permission.VIEW_SUSPICIOUS,
        # Operator-specific
        Permission.BAN_USERS,
        Permission.CREATE_ROOM,
        Permission.UPDATE_ROOM,
        Permission.DELETE_ROOM,
        Permission.SEND_ROOM_MESSAGE,
        Permission.APPROVE_WITHDRAWAL,
        Permission.CREATE_ANNOUNCEMENT,
        Permission.RESOLVE_SUSPICIOUS,
        Permission.VIEW_AUDIT_LOGS,
    },
    AdminRole.supervisor: {
        # All operator permissions
        Permission.VIEW_DASHBOARD,
        Permission.VIEW_METRICS,
        Permission.VIEW_USERS,
        Permission.VIEW_USER_DETAILS,
        Permission.BAN_USERS,
        Permission.VIEW_ROOMS,
        Permission.CREATE_ROOM,
        Permission.UPDATE_ROOM,
        Permission.DELETE_ROOM,
        Permission.SEND_ROOM_MESSAGE,
        Permission.VIEW_HANDS,
        Permission.EXPORT_HANDS,
        Permission.VIEW_DEPOSITS,
        Permission.VIEW_WITHDRAWALS,
        Permission.APPROVE_WITHDRAWAL,
        Permission.VIEW_WALLET,
        Permission.VIEW_ANNOUNCEMENTS,
        Permission.CREATE_ANNOUNCEMENT,
        Permission.VIEW_MAINTENANCE,
        Permission.VIEW_SUSPICIOUS,
        Permission.RESOLVE_SUSPICIOUS,
        Permission.VIEW_AUDIT_LOGS,
        # Supervisor-specific
        Permission.ADJUST_BALANCE,
        Permission.APPROVE_LARGE_WITHDRAWAL,
        Permission.FORCE_CLOSE_ROOM,
        Permission.BROADCAST_ANNOUNCEMENT,
        Permission.SCHEDULE_MAINTENANCE,
    },
    AdminRole.admin: {
        # All permissions
        Permission.VIEW_DASHBOARD,
        Permission.VIEW_METRICS,
        Permission.VIEW_USERS,
        Permission.VIEW_USER_DETAILS,
        Permission.BAN_USERS,
        Permission.ADJUST_BALANCE,
        Permission.APPROVE_LARGE_ADJUSTMENT,
        Permission.VIEW_ROOMS,
        Permission.CREATE_ROOM,
        Permission.UPDATE_ROOM,
        Permission.DELETE_ROOM,
        Permission.FORCE_CLOSE_ROOM,
        Permission.SEND_ROOM_MESSAGE,
        Permission.VIEW_HANDS,
        Permission.EXPORT_HANDS,
        Permission.VIEW_DEPOSITS,
        Permission.VIEW_WITHDRAWALS,
        Permission.APPROVE_WITHDRAWAL,
        Permission.APPROVE_LARGE_WITHDRAWAL,
        Permission.VIEW_WALLET,
        Permission.VIEW_ANNOUNCEMENTS,
        Permission.CREATE_ANNOUNCEMENT,
        Permission.BROADCAST_ANNOUNCEMENT,
        Permission.VIEW_MAINTENANCE,
        Permission.SCHEDULE_MAINTENANCE,
        Permission.ACTIVATE_MAINTENANCE,
        Permission.VIEW_SUSPICIOUS,
        Permission.RESOLVE_SUSPICIOUS,
        Permission.VIEW_AUDIT_LOGS,
        Permission.VIEW_ADMINS,
        Permission.CREATE_ADMIN,
        Permission.MODIFY_ADMIN,
        # Partner management
        Permission.VIEW_PARTNERS,
        Permission.CREATE_PARTNER,
        Permission.UPDATE_PARTNER,
        Permission.DELETE_PARTNER,
        # Settlement management
        Permission.VIEW_SETTLEMENTS,
        Permission.GENERATE_SETTLEMENT,
        Permission.APPROVE_SETTLEMENT,
        Permission.PAY_SETTLEMENT,
    },
    # 파트너 역할 - 자신의 데이터만 조회 가능
    AdminRole.partner: {
        Permission.VIEW_DASHBOARD,
        Permission.VIEW_SETTLEMENTS,  # 자신의 정산 내역만
    },
}


def has_permission(role: AdminRole, permission: Permission) -> bool:
    """Check if a role has a specific permission"""
    return permission in ROLE_PERMISSIONS.get(role, set())


def get_role_permissions(role: AdminRole) -> Set[Permission]:
    """Get all permissions for a role"""
    return ROLE_PERMISSIONS.get(role, set())
